grpc/tests/message_builder_test.cpp

```diff
@@ -39,121 +14,175 @@ bool release_n_verify(flatbuffers::grpc::MessageBuilder &mbb, const std::string
   return (monster->name()->str() == expected_name) && (monster->color() == color);
 }
 
-struct OwnedAllocator : public flatbuffers::DefaultAllocator {};
-
-struct TestHeapMessageBuilder : public flatbuffers::FlatBufferBuilder {
-  TestHeapMessageBuilder()
-    : flatbuffers::FlatBufferBuilder(2048, new OwnedAllocator(), true) {}
-};
-
-template <class Builder>
-struct BuilderTests {
-  static void empty_builder_movector_test() {
-    Builder b1;
-    size_t b1_size = b1.GetSize();
-    Builder b2(std::move(b1));
-    size_t b2_size = b2.GetSize();
-    test_assert(b1_size == 0);
-    test_assert(b1_size == b2_size);
+template <>
+struct BuilderReuseTests<flatbuffers::grpc::MessageBuilder> {
+  static void builder_reusable_after_release_message_test(TestSelector selector) {
+    if (!selector.count(REUSABLE_AFTER_RELEASE_MESSAGE)) {
+      return;
+    }
+
+    flatbuffers::grpc::MessageBuilder b1;
+    std::vector<flatbuffers::grpc::Message<Monster>> buffers;
+    for (int i = 0; i < 5; ++i) {
+      auto root_offset1 = populate1(b1);
+      b1.Finish(root_offset1);
+      buffers.push_back(b1.ReleaseMessage<Monster>());
+      TEST_ASSERT(verify(buffers[i], m1_name, m1_color));
+    }
   }
 
-  static void nonempty_builder_movector_test() {
-    Builder b1;
-    populate1(b1);
-    size_t b1_size = b1.GetSize();
-    Builder b2(std::move(b1));
-    test_assert(b1_size == b2.GetSize());
-    test_assert(0 == b1.GetSize());
+  static void builder_reusable_after_release_test(TestSelector selector) {
+    if (!selector.count(REUSABLE_AFTER_RELEASE)) {
+      return;
+    }
+
+    // FIXME: Populate-Release loop fails assert(GRPC_SLICE_IS_EMPTY(slice_)).
+
+    flatbuffers::grpc::MessageBuilder b1;
+    std::vector<flatbuffers::DetachedBuffer> buffers;
+    for (int i = 0; i < 5; ++i) {
+      auto root_offset1 = populate1(b1);
+      b1.Finish(root_offset1);
+      buffers.push_back(b1.Release());
+      TEST_ASSERT(verify(buffers[i], m1_name, m1_color));
+    }
   }
 
-  static void builder_movector_before_finish_test() {
-    Builder b1;
-    auto root_offset1 = populate1(b1);
-    Builder b2(std::move(b1));
-    b2.Finish(root_offset1);
-    test_assert(release_n_verify(b2, m1_name, m1_color));
-    test_assert(0 == b1.GetSize());
+  static void builder_reusable_after_releaseraw_test(TestSelector selector) {
+    if (!selector.count(REUSABLE_AFTER_RELEASE_RAW)) {
+      return;
+    }
+
+    flatbuffers::grpc::MessageBuilder b1;
+    for (int i = 0; i < 5; ++i) {
+      auto root_offset1 = populate1(b1);
+      b1.Finish(root_offset1);
+      size_t size, offset;
+      grpc_slice slice;
+      const uint8_t *buf = b1.ReleaseRaw(size, offset, slice);
```

Comment:
Do I understand correctly, that `slice` is owning the memory (refcount) and grpc_slice_unref destructs it...?

Reply:
yes, that's my understanding.

Reply:
More accurately, `grpc_slice_unref` destroys the data and refcount when the refcount hits zero. Otherwise, `slice` only refers to refcount.